thrift/controller.cc

```diff
@@ -57,30 +57,30 @@ future<> thrift_controller::do_start_server() {
     if (_server) {
         return make_ready_future<>();
     }
+    return seastar::async([this] {
```

Comment:
Threads are better for long-running code like compaction. Especially if there are many yield points that usually don't yield, like pushing data into a disk buffer, or a maybe_yield() point, since we can keep locals in registers.

Reply:
> since we can keep locals in registers

Why is this different for coroutines? In both cases you only have to spill registers into the stack frame / coroutine frame when you preempt, at least in principle, don't you?

The only in-principle problem with coroutines I see is that allocating a coroutine frame is more expensive than allocating a stack frame, so it's extra important to have small coroutines inlined, which is a bad thing.

Reply:
I think using a thread here is fine. Initializing a server is not on any hot paths, so we can afford to trade-off some performance (large initial allocation and slower context switches) for convenience.

Reply:
I mostly done it for simplicity and unification with similar code in cql and alternator. As I understand async stack is typically created once on startup and then freed when server is ready to accept connections.

Reply:
> Why is this different for coroutines? In both cases you only have to spill registers into the stack frame / coroutine frame when you preempt, at least in principle, don't you?

My limited understanding of our implementation hints that it may be due to `await_suspend` returning `void` which means that coroutine is suspended unconditionally and can be scheduled again. While for futures we run `need_preempt` before `yield`. Correct me if I am wrong here.

Reply:
> I think using a thread here is fine. Initializing a server is not on any hot paths, so we can afford to trade-off some performance (large initial allocation and slower context switches) for convenience.

Oh, I didn't meant to imply that there is any relevant performance effect here; of course it doesn't matter when used in the control plane.

I was thinking about the "heaviness" of the used abstraction. Coroutines as are convenient as threads, so since coroutines became available, the only good reason to use a thread is to add preemption to synchronous code. Seeing a thread sets an expectation that there might be some synchronous-but-preempting code being called underneath. Now, when someone wants to edit this code, they will have to check that they aren't breaking some contract by moving things outside a thread.

Reply:
Threads also allow for much better error handling and RAII semantics. But yeah, if those are not needed, then might as well use coroutines.

Reply:
> Threads also allow for much better error handling and RAII semantics.

What do you mean?

Reply:
In coroutines, you cannot use `co_await` in `catch {}` blocks and `constructors/destructors`. In a thread, you can use `get()` everywhere. Basically, a thread allows you to code just like you would in normal synchronous code, while in a coroutine you always have to be mindful of the fact that you are in a coroutine and work around its limitations.

Reply:
Also, re RAII in threads, one can use deferred_action / deferred_close and friends to gracefully call and await async methods like close/stop before destroying objects on the return or exception paths.